csharp/private/wrappers/dotnet.cc

```diff
@@ -1,26 +1,99 @@
+#include <fstream>
 #include <iostream>
 #include <sstream>
 #include <string>
 
 #ifdef _WIN32
+#include <direct.h>
 #include <errno.h>
 #include <process.h>
 #include <windows.h>
+#define getcwd _getcwd
 #else  // not _WIN32
 #include <stdlib.h>
 #include <unistd.h>
 #endif  // _WIN32
-
 #include "tools/cpp/runfiles/runfiles.h"
 
 using bazel::tools::cpp::runfiles::Runfiles;
+using SubstitutionMap = std::map<std::string, std::string>;
 
 std::string evprintf(std::string name, std::string path) {
   std::stringstream ss;
   ss << name << "=" << path;
   return ss.str();
 }
 
+/**
+ * Returns the full path of the current working directory.
+ *
+ * @return Full path of the current working directory. Empty if cannot be set.
+ */
+std::string PWD() {
+  char buffer[512];
+  char* location = getcwd(buffer, sizeof(buffer));
+  std::string result;
+  if (location) {
+    result = location;
+  }
+
+  return result;
+}
+
+/**
+ * Determines if a command argument is a parameter file.
+ *
+ * @param arg An argument passed to an action.
+ * @return true if the argument is a parameter file; false otherwise.
+ */
+bool IsParameterFile(std::string arg) { return arg[0] == '@'; }
+
+/**
+ * Replaces one or more keyword items in a string using the substitutions
+ * dictionary.
+ *
+ * @param arg An argument passed to an action.
+ * @param subst Substitutions to make when replacing keywords.
+ * @return The string representation of arg with substitutions.
+ */
+std::string Resolve(std::string arg, SubstitutionMap subst) {
+  for (const auto& kv : subst) {
+    size_t pos = arg.find(kv.first);
+    if (pos != std::string::npos)
+      arg.replace(pos, kv.first.length(), kv.second);
+  }
+
+  return arg;
+}
+
+/**
+ * Replaces keyword items in a parameters file using the substitutions
+ * dictionary.
+ *
+ * @param file The path to the parameters file.
+ * @param subst Substitutions to make when replacing keywords.
+ * @return true if the substitutions were made; false otherwise.
+ */
+bool ResolveParamsFile(std::string file, SubstitutionMap subst) {
```

Comment:
Yeah, so if we don't pass `/pathmap` in from Bazel we could just append it to argv rather than reaching into the file :)

There is the spawnv limit but since Windows paths are limited anyway it seems like it'd be guaranteed to fit along with the rest of our argv.

Reply:
I assume you mean using an environment variable as an action switch (`BAZEL_ACTION=ASSEMBLY_BUILD`), which would conditionally append `/pathmap` to the arguments?

Depending on the implementation of #10 & #72 , we may be making use of other commands of `dotnet` (publish / build /. etc), which wouldn't support the `/pathmap` argument. So to append that argument the wrapper would need to understand the command that is being executed.

I thought using variable substitutions would allow the least amount of friction in this case, as it wouldn't require the wrapper to understand the commands we executed, just how to do `${MY_VAR}` in a cross-platform way.

Reply:
As an addon to the above, the ideas that sort of pushed me towards using variable substitution were as follows:

- I don't want the `dotnet` wrapper to try and understand the commands I am executing
- I want to scrub away context `dotnet` requires in the build process (e.g. paths) 
- I want all the configuration about options/parameters/commands to remain in bazel

Reply:
> I assume you mean using an environment variable as an action switch (BAZEL_ACTION=ASSEMBLY_BUILD), which would conditionally append /pathmap to the arguments?

I'd prefer it if we just did it always for now, rather than conditionally.

> Depending on the implementation of #10 & #72 , we may be making use of other commands of dotnet (publish / build /. etc), which wouldn't support the /pathmap argument. So to append that argument, the wrapper would need to understand the command that is being executed.

Ok, let's cross that bridge when it comes to it. We can always have a single boolean switch that we pass to the wrapper if we need to, rather than the substitution thing. But let's not do anything prematurely.

> I thought using variable substitutions would allow the least amount of friction in this case, as it wouldn't require the wrapper to understand the commands we executed, just how to do ${MY_VAR} in a cross-platform way.

I think it's fine for the wrapper to understand this. It's all in the same repo and the wrapper is useless outside of this repo. Let's just keep it simple while we can.

Reply:
We are at that bridge with the current proposed implementation for compilation of resource files (see #98). The internals of that PR are using `dotnet build <csproj>` to compile resource files using the dotnet executable. This is the case I was describing above.